dex/src/main/java/io/crate/data/join/HashInnerJoinBatchIterator.java

```diff
@@ -0,0 +1,165 @@
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data.join;
+
+import com.google.common.collect.LinkedListMultimap;
+import com.google.common.collect.Multimap;
+import io.crate.data.ArrayRow;
+import io.crate.data.BatchIterator;
+import io.crate.data.Row;
+
+import javax.annotation.Nonnull;
+import java.util.Collection;
+import java.util.Iterator;
+import java.util.concurrent.CompletionStage;
+import java.util.function.Function;
+import java.util.function.Predicate;
+
+/**
+ * <pre>
+ *     Build Phase:
+ *     for (leftRow in left) {
+ *         calculate hash and put in Buffer (HashMap)
+ *     }
+ *     // All rows from left processed and inserted in the Buffer
+ *
+ *     Probe Phase:
+ *     // From now on we only iterate on the right
+ *     for (rightRow in right) {
+ *         if (hash(rightRow) found in Buffer {
+ *            for (row in matchedInBuffer) { // Handle duplicate values from left and hash-conflicts
+ *                emmit(combinedRow)
+ *            }
+ *         }
+ *     }
+ * </pre>
+ */
+class HashInnerJoinBatchIterator<L extends Row, R extends Row, C> extends JoinBatchIterator<L, R, C> {
+
+    private final Predicate<C> joinCondition;
+    private final Multimap<Integer, Object[]> buffer;
+    /**
+     * Used to avoid instantiating multiple times RowN in {@link #findMatchingRows()}
+     */
+    private final ArrayRow leftRow = new ArrayRow();
+    private final Function<L, Integer> hashForLeft;
+    private final Function<R, Integer> hashForRight;
+    private Iterator<Object[]> leftMatchingRowsIterator;
+
+    HashInnerJoinBatchIterator(BatchIterator<L> left,
+                               BatchIterator<R> right,
+                               ElementCombiner<L, R, C> combiner,
+                               Predicate<C> joinCondition,
+                               Function<L, Integer> hashForLeft,
+                               Function<R, Integer> hashForRight,
+                               int leftSize) {
+        super(left, right, combiner);
+        this.joinCondition = joinCondition;
+        this.hashForLeft = hashForLeft;
+        this.hashForRight = hashForRight;
+        this.buffer = LinkedListMultimap.create(leftSize);
+        this.activeIt = left;
+    }
+
+    @Override
+    public C currentElement() {
+        return combiner.currentElement();
+    }
+
+    @Override
+    public void moveToStart() {
+        left.moveToStart();
+        right.moveToStart();
+        activeIt = left;
+        buffer.clear();
+        leftMatchingRowsIterator = null;
+    }
+
+    @Override
+    public boolean moveNext() {
+        // Build HashMap from left relation
+        if (activeIt == left) {
+            while (left.moveNext()) {
+                Object[] currentRow = left.currentElement().materialize();
+                buffer.put(hashForLeft.apply(left.currentElement()), currentRow);
```

Comment:
why not use currentRow for creating the hash instead? would save some calls to the currentElement row..

Reply:
I'm not sure, because we somehow expose that the L, R (extends Row) are now (more concrete) Object[].
I can add a TODO when we glue them together